mlflow/llama_index/__init__.py

```diff
@@ -348,32 +354,158 @@ def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+
+    def _format_predict_input_query_and_retriever(self, data):
+        """Convert pyfunc input to a QueryBundle."""
+        data = _convert_llm_input_data(data)
+
+        if isinstance(data, str):
+            return QueryBundle.from_dict({"query_str": data})
+        elif isinstance(data, dict):
+            try:
+                return QueryBundle.from_dict(data)
+            except KeyError:
+                raise ValueError(
+                    "The input dictionary did not have the correct schema. It must support the "
+                    "supported in a llama_index QueryBundle: "
+                    "['query_str', 'image_path', 'custom_embedding_strs', 'embedding']"
+                )
+        elif isinstance(data, (list, np.ndarray)):
+            prediction_list = [self._format_predict_input_query_and_retriever(d) for d in data]
+        elif isinstance(data, pd.DataFrame):
+            prediction_list = [
+                self._format_predict_input_query_and_retriever(d)
+                for d in data.to_dict(orient="records")
+            ]
+        else:
+            raise ValueError(
+                f"Unsupported input type: {type(data)}. It must be one of "
+                "[str, dict, list, numpy.ndarray, pandas.DataFrame]"
+            )
+
+        if len(prediction_list) == 1:
+            return prediction_list[0]
+
+        return prediction_list
+
+    def _safe_convert_to_chat_messages(self, data) -> Union[List[ChatMessage], str]:
+        # NB: Chat messages that have been converted to a string via str() are formatted as
+        # "`role`: `content`", for example "system: You are a helpful bot".
+        if isinstance(data, (list, tuple)) and all(":" in message for message in data):
+            return [ChatMessage.from_str(*(message.split(": ", 1)[::-1])) for message in data]
+
+        return data
+
+    def _format_predict_input_chat(self, data):
+        """
+        Chat engines can have a variety of signatures. The two primary standards are:
+        1. chat(str, Sequence[ChatMessage])
+        2. chat(Sequence[ChatMessage])
+
+        There are not consistent naming conventions for either of these standards, so
+        we'll need to leverage positional arguments.
+        """
+        data = _convert_llm_input_data(data)
+
+        if isinstance(data, str):
+            return data
+        elif isinstance(data, dict):
+            if len(data) == 1:
+                # Assume that this value should be treated as a positional argument
+                payload = next(value for value in data.values())
+                return self._safe_convert_to_chat_messages(payload)
+
+            elif len(data) > 1:
+                # Assume that any kwarg value that's a sequence of strings with a colon
+                # should be converted to a list of ChatMessage objects.
+                return {k: self._safe_convert_to_chat_messages(v) for k, v in data.items()}
```

Comment:
What is the corresponding LlamaIndex invocation signature for this case? What will be the key for each chat message?

Reply:
There are several signatures, both in terms of number of parameters and the names of the parameters. There is also often a `**kwargs` in the signature. Some examples:
* `def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:`
* `def chat(self, message: str, chat_history: Optional[List[ChatMessage]] = None) -> AgentChatResponse:`
* `def chat(self, query: str, chat_history: Optional[List[ChatMessage]] = None):`

[Search term in llama_index](https://github.com/search?q=repo%3Arun-llama%2Fllama_index+%22def+chat%22+-path%3Alegacy&type=code)

Here's the thought process
* All signatures accept 1 positional argument, so if the dict has `len == 1`, it should be treated as positional
* All signatures require a message history formatted as `List[ChatMessage]`. The user must pass a list of string `[str(ChatMessage(role='user', content='xyz'))]` (llama_index implements a `__repr__` in `ChatMessage`). Then, we parse this via `_safe_convert_to_chat_messages`.

Would appreciate any suggestions on how to make this more robust - it's a start but is not very dynamic 🙂

Reply:
Thanks for the explanation, @michael-berk !!

> All signatures accept 1 positional argument, so if the dict has len == 1, it should be treated as positional

Hmm, the `message: str` and `query: str` is not the stringified chat message, but a single question? So the input will look like:

```
model.predict({
    "message": "What is MLflow?"
    "chat_history": [
        {
            "role": "system",
            "content": "xxx",
        },
        {
            "role": "user",
            "content": "yyy",
        }
    ]
```

Also here `chat_history` is passed as list of dictionary, not a list of `ChatMessage`. We need to support both, because in model serving, the request payload is always JSON serialized.

> The user must pass a list of string [str(ChatMessage(role='user', content='xyz'))]

I don't think we can ask that favor to users.... the same input example has to work for  models loaded by `mlflow.llamaindex.load_model()` and `mlflow.pyfunc.load_model()`. Also `str` conversion is not reliable in general, like they (llamaindex) can change `__repr__` implementation to whatever convenient, e.g. omits some field.

My suggestion here is to process the `query`/`message` and other messages separately. What do you think??🤔

```
   messages = []
   if user_query := data.get("query") or data.get("message"):
        messages.append(ChatMessage(role="user", content=user_query)
   if history := data.get("chat_history"):
       messages.extend([ChatMessage(**msg) for msg in history])
```

Reply:
Nice! Resolved in [this commit](https://github.com/mlflow/mlflow/pull/11674/commits/99478107bb4a3c488794d8139223c20718112d17)

I've added all kwargs that currently exist in llama_index for both `query` and `message_history`.